### ISSUE TITLE
contextEstablished from Exchange to Message.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#487]
Currently for established an observation : 
1) we add the observation to the store without endpointContext.
2) then when endpointContext is established we set the endpointContext to the store.
3) we send the request.

I think this is good point to add the observation into the store before to send the request.
But we could do that just after the context was established. This could avoid the double access to the store.
Another point instead of having the event on the exchange, I think this should be clearer to have it on the request/message.

like this (in `BaseMatcher.registerObserver()`) :
```java
request.addMessageObserver(new MessageObserverAdapter() {
        public void onContextEstablished(EndpointContext context ){
               observationStore.add(new Observation(request, context));
        }

	@Override
	public void onCancel() {
		failed();
	}

	@Override
	protected void failed() {
		observationStore.remove(request.getToken());
		exchangeStore.releaseToken(idByToken);
	}
});
```

**Comment by boaks** [comment]
Just to mention: when i made the endpoint context, I also felt, that this should be adjusted. But, even if I have no precise information, I also think, that defining the "uniqueness of the token" may have also influence on that, and so I postponed this change. But, though leshan is the main user of that API, it's on you simon, if you want to change it now and potentially again with the token finalization.

**Comment by boaks** [comment]
Included in PR #551

**Comment by sbernard31** [comment]
Integrated in master (see #590)
